cpp/include/raft/linalg/reduce.cuh

```diff
@@ -75,6 +78,98 @@ void reduce(OutType* dots,
     dots, data, D, N, init, rowMajor, alongRows, stream, inplace, main_op, reduce_op, final_op);
 }
 
+/**
+ * @defgroup reduction Reduction Along Requested Dimension
+ * @{
+ */
+
+/**
+ * @brief Compute reduction of the input matrix along the requested dimension
+ *
+ * @tparam InElementType the input data-type of underlying raft::matrix_view
+ * @tparam LayoutPolicy The layout of Input/Output (row or col major)
+ * @tparam OutElementType the output data-type of underlying raft::matrix_view and reduction
+ * @tparam IndexType Integer type used to for addressing
+ * @tparam MainLambda Unary lambda applied while acculumation (eg: L1 or L2 norm)
+ * It must be a 'callable' supporting the following input and output:
+ * <pre>OutType (*MainLambda)(InType, IdxType);</pre>
+ * @tparam ReduceLambda Binary lambda applied for reduction (eg: addition(+) for L2 norm)
+ * It must be a 'callable' supporting the following input and output:
+ * <pre>OutType (*ReduceLambda)(OutType);</pre>
+ * @tparam FinalLambda the final lambda applied before STG (eg: Sqrt for L2 norm)
+ * It must be a 'callable' supporting the following input and output:
+ * <pre>OutType (*FinalLambda)(OutType);</pre>
+ * @param handle raft::handle_t
+ * @param dots Output of type raft::device_matrix_view
+ * @param data Input of type raft::device_matrix_view
+ * @param init initial value to use for the reduction
+ * @param apply whether to reduce along rows or along columns (using raft::linalg::Apply)
+ * @param main_op elementwise operation to apply before reduction
```

Comment:
when we accept lambdas like this, can we make use of the term `fused` somewhere in these descriptions so that we can let the user know the benefits of this element-wise operation up front? We have an element-wise primitive so why are they better off using this one? (the answer: because it fuses the element-wise op into the kernel before performing the reduction). 

If a lambda op is not fused and needs to be performed as a separate operation, we'd be better off removing it as an argument because it would save the additional template parameter which is non-trivial to specialize generically.